migrator/migrations/m_196_to_m_197_set_poduid_where_null/migration_impl.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/stackrox/rox/migrator/types"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/postgres/pgutils"
+	"github.com/stackrox/rox/pkg/sac"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// The purpose of this migration is to set the value of PodUid in the listening_endpoints table where
+// it is null and it is possible to give it a value.
+func migrate(database *types.Databases) error {
+	ctx := sac.WithAllAccess(context.Background())
+	pgutils.CreateTableFromModel(ctx, database.GormDB, podSchema.CreateTablePodsStmt)
+	pgutils.CreateTableFromModel(ctx, database.GormDB, listeningEndpointsSchema.CreateTableListeningEndpointsStmt)
+	pgutils.CreateTableFromModel(ctx, database.GormDB, processIndicatorSchema.CreateTableProcessIndicatorsStmt)
+
+	return updateGlobalScope(ctx, database)
+}
+
+// Walk through the process_indicators table and get a map where the key is the id and the value is the
+// PodUid. It doesn't do this for every process indicator, but only for the ones where it could possibly
+// be used to set PodUid in the listening endpoints table.
+func getProcessIndicatorPodUIDMap(ctx context.Context, processIndicatorStore processIndicatorDatastore.Store, processIndicatorIds map[string]bool) (map[string]string, error) {
+	podUIDMap := make(map[string]string)
+
+	err := processIndicatorStore.Walk(ctx,
+		func(processIndicator *storage.ProcessIndicator) error {
+			_, exists := processIndicatorIds[processIndicator.Id]
+			if exists {
+				podUIDMap[processIndicator.Id] = processIndicator.GetPodUid()
+				delete(processIndicatorIds, processIndicator.Id)
+			}
+			return nil
+		})
+
+	processIndicatorIds = make(map[string]bool)
+
+	return podUIDMap, err
+}
+
+// Find out the process indicator ids of plops that could possibly be updated with a poduid from the
+// process_indicators table. This way we don't have to store all id, poduid pairs from the
+// process_indicators table. Go doesn't have sets so use a map where the value is always true instead.
+func getProcessIndicatorIdsOfInterest(ctx context.Context, plopStore plopDatastore.Store) (map[string]bool, error) {
+	processIndicatorIds := make(map[string]bool)
+
+	err := plopStore.Walk(ctx,
+		func(plop *storage.ProcessListeningOnPortStorage) error {
+			if plop.GetPodUid() == "" && plop.GetProcess() == nil {
+				processIndicatorIds[plop.GetProcessIndicatorId()] = true
+			}
+			return nil
+		})
+
+	return processIndicatorIds, err
+}
+
+// Walks through the listening_endpoints table and sets the value of PodUid where it is null using a
+// map obtained from the process_indicators table, where the key is a processindicatorid and the value
+// is a PodUid.
+func setPodUidsUsingProcessIndicators(ctx context.Context, plopStore plopDatastore.Store, podUIDMap map[string]string, batchSize int) error {
+	plops := make([]*storage.ProcessListeningOnPortStorage, batchSize)
```

Comment:
The plops slice used for the batch is handled a bit differently here than in other migrations. I don't know if it is better or worse. I set the length to the batchSize. The reason I did that is I don't like append, which results in the array being copied if the capacity is reached. That is expensive in both time and memory. However, the way the code is currently it doesn't clear out the memory used by the plops when the batch size is reached. That could be fixed.